[ORCA-155] Implement basic Nextflow Tower service

BrunoGrandePhD opened the pull request:
I wanted to use the new base classes to implement a new service, so I chose Nextflow Tower. This implementation is very basic, but it serves to demonstrate the utility of the base classes. 

I realized that I can probably parametrize the tests as well for different services to avoid duplication. I'll look into that next. 

- [x] Implement tests for Nextflow Tower

Commit:
Implement Nextflow Tower service (w/o tests)

Commit:
Fix missing import

BrunoGrandePhD commented:
I'll mark this as ready for review to get feedback on the existing PR, but I plan on adding tests before merging.

BWMac commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+    """
+
+    auth_token: str
+    api_endpoint: str
```
In `sagetasks` we had a set list of API endpoints in a dictionary similar to the `SevenBridges` implementation in this repo. Is the reason to not include that here to allow for non-sage use cases?

BrunoGrandePhD replied:
In the case of SevenBridges, there are no other possible options for `api_endpoint`. On the other hand, there are multiple deployments of Nextflow Tower. So, you are correct in saying that this decision is to allow for non-Sage use cases.

BWMac commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+            )
+            raise ValueError(message)
+
+    def request(self, method: str, path: str, **kwargs) -> requests.Response:
```
I like having `request` as an option because it gives you full access to the tower API functionality, but just want to confirm that more convenient methods like `launch_workflow` from `sagetasks` will be a part of future work

BrunoGrandePhD replied:
That's correct! We can add the convenience methods like `launch_workflow()` as we add the correponding function in `NextflowTowerOps`.

thomasyu888 commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+        """Iterate through pages of results for a given request.
+
+        See ``TowerClient.request`` for argument definitions.
+
```
Nit: Add documentation all the way through about **kwargs being `requests.requests(**kwargs)`

BrunoGrandePhD replied:
The docstring already states:

>  See ``TowerClient.request`` for argument definitions.

And `TowerClient.request` defines the following:

> **kwargs: Additional named arguments passed through to requests.request().

Is there something that you want to add? To clarify, my intent here is to avoid duplication of documentation.

thomasyu888 replied:
100% just glossed over that.  Thanks for pointing that out!

thomasyu888 commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+        Returns:
+            The cumulative list of items from all pages.
+        """
+        self.update_kwarg(kwargs, "params", "max", 50)
```
I understand the reason for these two parameters and actually like the way you've implemented it, but what is your reason for not including `max` and `offset` as an argument (`def request_paged(...., max=50...)`) and as a kwarg?

BrunoGrandePhD replied:
For consistency with the other `request` methods and for simplicity. For example, if I add a `max` parameter, I now need to handle (and document) the case where a user specifies `max` in the parameter and in the `params` dictionary. The user already has a way to provide this value, so I don't see the need to add another way, especially since I don't expect users to edit this value very often.

thomasyu888 replied:
+1

thomasyu888 commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+        See ``TowerClient.request`` for argument definitions.
+
+        Raises:
+            ValueError: If the 'params' keyword argument
```
Nit: Raise HTTPError?

BrunoGrandePhD replied:
Good catch! I reviewed all of the docstrings and made sure that the `Raises` section is consistent.

thomasyu888 commented on src/orca/services/nextflowtower/config.py:
```diff
@@ -0,0 +1,55 @@
+
+    api_endpoint: Optional[str] = None
+    auth_token: Optional[str] = None
+    workspace: Optional[int] = None
```
Should this be `Optional[str]`?

BrunoGrandePhD replied:
That's the type of the organization and workspace IDs according to the Tower OpenAPI [specification](https://api.tower.nf/openapi#put-/orgs/-orgId-/workspaces/-workspaceId-). 

<img width="818" alt="image" src="https://user-images.githubusercontent.com/740725/223850037-48035d1d-51b5-4ca6-88a7-09a42cb1a71e.png">

BrunoGrandePhD replied:
I just noticed your comment [here](https://github.com/Sage-Bionetworks-Workflows/py-orca/pull/13#pullrequestreview-1331380427). I realize the confusion. While the docstring specifies that this value should be the numerical workspace ID, the attribute name (no `_id`) doesn't imply that. I agree that it's convenient to specify workspace names. The Tower CLI supports this with this format: `<org-name>/<workspace-name>` (similar to project names in SevenBridges). I'll add support for both methods.

thomasyu888 reviewed (changes requested):
Awesome work, but the below failed for me.

```
from orca.services.nextflowtower.client_factory import (
    NextflowTowerConfig,
)

from orca.services.nextflowtower.ops import (
    NextflowTowerOps,
)
auth_token = ""

# I modified the type in Config as well to support this workspace.
config = NextflowTowerConfig(
    api_endpoint="https://tower.sagebionetworks.org/api",
    auth_token=auth_token,
    workspace="genie-bpc-project"
)

ops = NextflowTowerOps(
    config=config
)

ops.client
```

This caused this error:

```
Traceback (most recent call last):
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 79, in test_client
    cls.test_client_request(client)
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client_factory.py", line 62, in test_client_request
    client.get_user_info()
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 128, in get_user_info
    return self.request_json("get", path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 78, in request_json
    response = self.request(method, path, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 58, in request
    raise ValueError(message)
ValueError: Method (get) not among valid options ({'PUT', 'GET', 'POST', 'DELETE'}).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/base/ops.py", line 39, in client
    return factory.get_client(test=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 100, in get_client
    self.test_client(client)
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 82, in test_client
    raise ClientRequestError(message) from error
orca.errors.ClientRequestError: Authenticated test request failed using the client.
>>> ops.client
Traceback (most recent call last):
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 79, in test_client
    cls.test_client_request(client)
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client_factory.py", line 62, in test_client_request
    client.get_user_info()
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 128, in get_user_info
    return self.request_json("get", path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 78, in request_json
    response = self.request(method, path, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/nextflowtower/client.py", line 58, in request
    raise ValueError(message)
ValueError: Method (get) not among valid options ({'PUT', 'GET', 'POST', 'DELETE'}).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/base/ops.py", line 39, in client
    return factory.get_client(test=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 100, in get_client
    self.test_client(client)
  File "/Users/tyu/sage/py-orca/src/orca/services/base/client_factory.py", line 82, in test_client
    raise ClientRequestError(message) from error
orca.errors.ClientRequestError: Authenticated test request failed using the client
```

thomasyu888 commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+            Current user
+        """
+        path = "/user-info"
+        return self.request_json("get", path)
```
I think this needs to be uppercase? "GET" (Or see my previous comment)

BrunoGrandePhD replied:
Yeah, that's a bug! I added a convenience `get()` method to minimize the risk for this error.

thomasyu888 commented on src/orca/services/nextflowtower/client.py:
```diff
@@ -0,0 +1,128 @@
+        Returns:
+            The raw Response object to allow for special handling
+        """
+        valid_methods = {"GET", "PUT", "POST", "DELETE"}
```
Nit: you could uppercase the `method` here to allow for lower case methods as parameters in functions.

BrunoGrandePhD replied:
I considered that, but I prefer the `get()` convenience method, which is more elegant:

```python
return self.request_json("GET", path)
# versus
return self.get(path)
```

thomasyu888 replied:
👍🏼

Commit:
Address Tom's PR comments

Commit:
Switch from workspace ID to fully-qualified name

BrunoGrandePhD commented:
@thomasyu888: Your feedback was very valuable. I've updated the config to expect the fully-qualified workspace name (_e.g._ `sage-bionetworks/genie-bpc-project`), which are more user-friendly than workspace IDs. I have added a cached property to the Ops class to retrieve the workspace ID once when requested since most workspace-related requests require the ID and not the name. 

The following code should now work for you: 

```python
from orca.services.nextflowtower import NextflowTowerConfig, NextflowTowerOps

auth_token = "REDACTED"

config = NextflowTowerConfig(
    api_endpoint="https://tower.sagebionetworks.org/api",
    auth_token=auth_token,
    workspace="sage-bionetworks/genie-bpc-project",
)

ops = NextflowTowerOps(config)
ops.client
```

thomasyu888 reviewed (commented):
Excellent update @BrunoGrandePhD .  Just a question, when I run this:

```
from orca.services.nextflowtower import (
    NextflowTowerConfig,
    NextflowTowerOps
)
auth_token = ""
config = NextflowTowerConfig(
    api_endpoint="https://tower.sagebionetworks.org/api",
    auth_token=auth_token,
    workspace="sage-bionetworks/genie-bpc-project"
)
ops = NextflowTowerOps(config=config)

# response = ops.client.request("GET", "/pipelines")
# response.content
# response = ops.client.request("GET", "/workflow/random-name")
# response.content

# This shows 0 pipelines even though I see 3 pipelines in the UI
ops.client.get("/pipelines")
# This works
ops.client.get("/workflow/random-name")
# This works
ops.workspace_id
# 5355285966491
```

Is this a permission issue with the access token that I don't see any pipelines even though I see them in the UI?

Commit:
Handle unset workspace attribute

BrunoGrandePhD commented:
Good catch! That's happening because `NextflowTowerClient` will not automatically populate the `workspaceId` query parameter. This is something that needs to be done explicitly (see below). Otherwise, you are presumably pulling the list of pipelines in your user workspace, which we disabled on the Sage Tower instance. 

My plan is to make the `NextflowTowerOps` class workspace-aware similar to how the `SevenBridgesOps` class is project-aware. I opted against making the `NextflowTowerClient` workspace-aware for consistency with the SevenBridges client. That said, we can revisit this. Ideally, most common operations will be implemented in `NextflowTowerOps`, obviating the need to explicitly set the workspace. 

```python
ops.client.get("/pipelines", params={"workspaceId": ops.workspace_id})
```

thomasyu888 reviewed (commented):
🦖 Thanks @BrunoGrandePhD !  That makes sense and your command worked!

```
ops.client.get("/pipelines", params={"workspaceId": ops.workspace_id})
```

We should have a short meeting about adding tickets to this EPIC [Nextflow Tower Hooks EPIC](https://sagebionetworks.jira.com/browse/ORCA-144) .  This way we can start distributing some of the work since you've put together this beautiful framework.

Before I approve - I noticed you have a task to add more tests - is that still true?  Or... do you think we should add them in as we add in functions in `ops`?

Commit:
add initial base class tests

Commit:
add notes/to dos, add add. tests for config

Commit:
Refactor base tests

Commit:
Merge pull request #14 from Sage-Bionetworks-Workflows/rxu17/ORCA-155/tower-service

[ORCA-155] Adding tests to base tower service class

Commit:
Make sure that all services are tested

Commit:
Fill in test coverage for `nextflowtower`

Commit:
Add connection URI for Tower service

thomasyu888 commented on tests/services/nextflowtower/responses.py:
```diff
@@ -0,0 +1,65 @@
+https://tower-dev.sagebionetworks.org/api/openapi
+"""
+
+get_user_info = {
```
What do you think about these making these as fixtures and yielding them to act as setup and teardown?

BrunoGrandePhD replied:
I'm not a fan of having to include another fixture in every test function that needs to use a response, but if you're worried about ensuring that tests are independent by preventing the modification of mutable shared dictionaries, then there's not much choice. I addressed this here: https://github.com/Sage-Bionetworks-Workflows/py-orca/pull/13/commits/5e3a3bee36c7d258f1e6b1d5e97301f0579654e1

BWMac commented on src/orca/services/nextflowtower/config.py:
```diff
@@ -0,0 +1,81 @@
+
+    connection_env_var = "NEXTFLOWTOWER_CONNECTION_URI"
+
+    @validator("workspace")
```
TIL the validator decorator. Very cool.

BWMac reviewed (approved):
LGTM. I used the following code and everything worked as expected:
```
#import
from orca.services.nextflowtower.client_factory import NextflowTowerClientFactory
from orca.services.nextflowtower.config import NextflowTowerConfig
from orca.services.nextflowtower.ops import NextflowTowerOps

#confiugration variables
auth_token = os.environ.get("TOWER_ACCESS_TOKEN")
api_endpoint = "https://tower.sagebionetworks.org/api"
workspace = "Sage-Bionetworks/example-project"

#set config
config = NextflowTowerConfig(api_endpoint=api_endpoint, auth_token=auth_token,workspace=workspace)

#create factory
factory = NextflowTowerClientFactory(config=config)

#create client
client = factory.create_client()

#test client functionality
info = client.get_user_info()
client.get_user_workspaces_and_orgs(info["id"])
client.list_user_workspaces()

#create ops
ops = NextflowTowerOps(config=config)

#test ops properties
ops.workspace_id
ops.workspace
```

thomasyu888 reviewed (approved):
💯 awesome! Just a minor comment, but going to pre-approve.

Commit:
Make response dicts immutable through fixture
